### PR TITLE
Simpler stream construction

### PR DESCRIPTION
**Pull request by sonewman**
This is a PR to support suggested functionality discussed in issue https://github.com/iojs/readable-stream/issues/102 to be discussed at WG Meeting: https://github.com/iojs/readable-stream/issues/106

- [x] Create prototype with ability to set *required* methods through stream construction
- [ ] Discuss / review implementation with WG (@iojs/streams) members.
- [ ] Finalise on best implementation
- [ ] Add necessary unit tests to prove functionality

*(Subsequent changes will likely be rebased into one commit)*

**Please review - Do Not Merge**

### CODE REVIEW

**Comment by chrisdickinson** [comment]
One thing that will be a prerequisite before this gets merged is to put in the work to reverse the flow of readable-stream into io.js.

File: lib/_stream_readable.js [hunk 1]
```diff
@@ -121,6 +121,9 @@ function Readable(options) {
   // legacy
   this.readable = true;
 
+  if (options && util.isFunction(options.read))
+    this._read = options.read;
+
```
Comment (chrisdickinson) [inline lib/_stream_readable.js]: Hm, I'm -0 on adding this if it's a shorthand for `_read = fn` – accepting functions into the constructor is a natural jumping off point for the other changes we'd like to see with streams, and if we don't take that opportunity when we add this functionality we'll have lost the chance to do so in the future (this roughly echoes my position from the other issue.)

**Comment by sonewman** [comment]
That's fine, it was more like a *rough* here was suggesting, let's discuss. I don't think it is necessarily the final solution.

File: lib/_stream_readable.js (hunk 1 above)
Comment (sonewman) [inline lib/_stream_readable.js]: Ahh so you were suggest jump straight to 'start' & 'pull'?

**Comment by sonewman** [comment]
It's definitely best if we do this properly so we are all in agreement, I am just keen to prompt discussion.

**Comment by chrisdickinson** [comment]
Per the WG meeting: everyone was pretty much +1 on this going in if the keys for _transform and _write don't conflict with WHATWG-style keys.

**Comment by sonewman** [comment]
@chrisdickinson cool, well I'll wait for @domenic to come back with some feedback before writing some *simple* tests.

As you know I have raised some questions in (https://github.com/iojs/readable-stream/issues/102) maybe they were your initial concerns before?

**Comment by domenic** [comment]
OK, so basically the semantics are compatible in all cases. (Except readable streams, but we dodge that by using `read` here which doesn't conflict with WHATWG streams `start` and `pull`.) This harmoniousness includes writev and transform/flush, which aren't finalized for WHATWG streams yet but I am pretty sure what they'll look like. All the important things are there: the same hooks are async in the same way, and roughly the same data is passed in.

In general the parameters don't match up, even ignoring the promise/callback issues. So people won't be able to transparently swap out `new stream.Readable({...})` with `new ReadableStream({ ... })`. But we knew that would happen.

So, +1 to proceed.

**Comment by chrisdickinson** [comment]
@domenic Awesome, thanks for looking into this.

@sonewman So, in the interest of expedience, you might re-open this PR over on iojs/io.js, and cc me there. We can merge it in there, then use the existing slurping tools to bring it into this repo. Great work!

**Comment by sonewman** [comment]
@domenic @chrisdickinson awesome things for the update, I will knock up some tests and raised as a PR in iojs/io.js also.

**Comment by sonewman** [comment]
This is in core, so I am going to close

**Pull request closed without merging**